CI still fails even if a rerun of a UI test passes

PR #37785  added some new logic to keep track of acceptance test scenarios that fail, and report them all at the end.

But now if a UI test fails, and it is automatically rerun, and the rerun passes then the test pipeline still fails. For example:
https://drone.owncloud.com/owncloud/core/26336/105/15
```
--- Failed scenarios:

    /drone/src/tests/acceptance/features/webUICreateDelete/createFoldersEdgeCases.feature:53

35 scenarios (34 passed, 1 failed)
371 steps (367 passed, 1 failed, 3 skipped)
21m53.22s (25.35Mb)
webUI test run failed with exit status: 1
Rerun failed scenario: /drone/src/tests/acceptance/features/webUICreateDelete/createFoldersEdgeCases.feature:53
@webUI @insulated @disablePreviews
Feature: create folder
  As a user
  I want to create folders
  So that I can organise my data structure

  Background:                                                                              # /drone/src/tests/acceptance/features/webUICreateDelete/createFoldersEdgeCases.feature:7
    Given user "Alice" has been created with default attributes and without skeleton files # FeatureContext::userHasBeenCreatedWithDefaultAttributesAndWithoutSkeletonFiles()
    And user "Alice" has logged in using the webUI                                         # WebUILoginContext::theUserHasLoggedInUsingTheWebUI()
    And the user has browsed to the files page                                             # WebUIFilesContext::theUserBrowsesToTheFilesPage()

SCENARIO RESULT: (pass)
  @smokeTest @skipOnLDAP
  Scenario Outline: Create a sub-folder inside an existing folder with problematic name # /drone/src/tests/acceptance/features/webUICreateDelete/createFoldersEdgeCases.feature:43
    Given user "Alice" has created folder <folder>                                      # FeatureContext::userHasCreatedFolder()
    And the user has reloaded the current page of the webUI                             # WebUIGeneralContext::theUserReloadsTheCurrentPageOfTheWebUI()
    When the user opens folder <folder> using the webUI                                 # WebUIFilesContext::theUserOpensFolderNamedUsingTheWebUI()
    And the user creates a folder with the name "sub-folder" using the webUI            # WebUIFilesContext::theUserCreatesAFolderUsingTheWebUI()
    Then folder "sub-folder" should be listed on the webUI                              # WebUIFilesContext::fileFolderShouldBeListedOnTheWebUI()
    When the user reloads the current page of the webUI                                 # WebUIGeneralContext::theUserReloadsTheCurrentPageOfTheWebUI()
    Then folder "sub-folder" should be listed on the webUI                              # WebUIFilesContext::fileFolderShouldBeListedOnTheWebUI()

    Examples:
      | folder |
      | "0"    |

1 scenario (1 passed)
10 steps (10 passed)
0m34.81s (22.70Mb)
runsh: Total 35 scenarios (34 passed, 1 failed)
runsh: Exit code of main run: 0
runsh: Total unexpected failed scenarios throughout the test run:
webUICreateDelete/createFoldersEdgeCases.feature:53
runsh: There were no unexpected success.
make: *** [test-acceptance-webui] Error 1
Makefile:215: recipe for target 'test-acceptance-webui' failed
```

`tests/acceptance/run.sh` needs more logic so that it understands if all the rerun scenarios have actually passed.

Happened in files_primary_s3 last night:
https://drone.owncloud.com/owncloud/files_primary_s3/1833/23/19
```
runsh: Total 52 scenarios (51 passed, 1 failed)
runsh: Exit code of main run: 0
runsh: Total unexpected failed scenarios throughout the test run:
webUIManageUsersGroups/deleteUsers.feature:37
runsh: There were no unexpected success.
Makefile:133: recipe for target 'test-acceptance-core-webui' failed
make: *** [test-acceptance-core-webui] Error 1
```

Actually `webUIManageUsersGroups/deleteUsers.feature:37` was rerun and passed, but the CI pipeline end with fail.

And in encryption last night:
https://drone.owncloud.com/owncloud/encryption/1383/172/17
```
--- Failed scenarios:

    /var/www/owncloud/testrunner/tests/acceptance/features/webUIPersonalSettings/personalGeneralSettings.feature:64

29 scenarios (28 passed, 1 failed)
190 steps (188 passed, 1 failed, 1 skipped)
30m19.56s (23.39Mb)
webUI test run failed with exit status: 1
Rerun failed scenario: /var/www/owncloud/testrunner/tests/acceptance/features/webUIPersonalSettings/personalGeneralSettings.feature:64
@webUI @insulated @disablePreviews
Feature: personal general settings
  As a user
  I want to change the ownCloud User Interface to my preferred settings
  So that I can personalise the User Interface

  Background:                                                                              # /var/www/owncloud/testrunner/tests/acceptance/features/webUIPersonalSettings/personalGeneralSettings.feature:7
    Given user "Alice" has been created with default attributes and without skeleton files # FeatureContext::userHasBeenCreatedWithDefaultAttributesAndWithoutSkeletonFiles()
    And user "Alice" has logged in using the webUI                                         # WebUILoginContext::theUserHasLoggedInUsingTheWebUI()
    And the user has browsed to the personal general settings page                         # WebUIPersonalGeneralSettingsContext::theUserBrowsesToThePersonalGeneralSettingsPage()

  Scenario: User uploads new profile picture                                                 # /var/www/owncloud/testrunner/tests/acceptance/features/webUIPersonalSettings/personalGeneralSettings.feature:64
    Given user "Alice" has uploaded file "filesForUpload/testavatar.jpg" to "/testimage.jpg" # FeatureContext::userHasUploadedAFileTo()
    And the user has deleted any existing profile picture                                    # WebUIPersonalGeneralSettingsContext::theUserHasDeletedAnyExistingProfilePicture()
    When the user uploads "testavatar.png" as a new profile picture using the webUI          # WebUIPersonalGeneralSettingsContext::theUserUploadsAsANewProfilePicture()
    Then the preview of the profile picture should be shown on the webUI                     # WebUIPersonalGeneralSettingsContext::thePreviewOfTheProfilePictureShouldBeShownOnTheWebui()
SCENARIO RESULT: (pass)

1 scenario (1 passed)
7 steps (7 passed)
1m4.40s (22.06Mb)
runsh: Total 29 scenarios (28 passed, 1 failed)
runsh: Exit code of main run: 0
runsh: Total unexpected failed scenarios throughout the test run:
webUIPersonalSettings/personalGeneralSettings.feature:64
runsh: There were no unexpected success.
Makefile:142: recipe for target 'test-acceptance-core-webui' failed
make: *** [test-acceptance-core-webui] Error 1
```

And again in encryption last night:
https://drone.owncloud.com/owncloud/encryption/1383/188/17
```
--- Failed scenarios:

    /var/www/owncloud/testrunner/tests/acceptance/features/webUISharingPublic2/shareByPublicLink.feature:151

33 scenarios (32 passed, 1 failed)
328 steps (326 passed, 1 failed, 1 skipped)
30m48.72s (27.47Mb)
webUI test run failed with exit status: 1
Rerun failed scenario: /var/www/owncloud/testrunner/tests/acceptance/features/webUISharingPublic2/shareByPublicLink.feature:151
@webUI @insulated @disablePreviews @mailhog @public_link_share-feature-required @files_sharing-app-required
Feature: Share by public link
  As a user
  I want to share files through a publicly accessible link
  So that users who do not have an account on my ownCloud server can access them
  
  As an admin
  I want to limit the ability of a user to share files/folders through a publicly accessible link
  So that public sharing is limited according to organization policy

  Background:                                                                              # /var/www/owncloud/testrunner/tests/acceptance/features/webUISharingPublic2/shareByPublicLink.feature:11
    Given user "Alice" has been created with default attributes and without skeleton files # FeatureContext::userHasBeenCreatedWithDefaultAttributesAndWithoutSkeletonFiles()

  Scenario: user removes the public link of a file                                                                        # /var/www/owncloud/testrunner/tests/acceptance/features/webUISharingPublic2/shareByPublicLink.feature:151
    Given user "Alice" has uploaded file "filesForUpload/lorem.txt" to "/lorem.txt"                                       # FeatureContext::userHasUploadedAFileTo()
    And user "Alice" has created a public link share of file "/lorem.txt"                                                 # FeatureContext::userHasCreatedAPublicLinkShareOf()
    And user "Alice" has logged in using the webUI                                                                        # WebUILoginContext::theUserHasLoggedInUsingTheWebUI()
INFORMATION: timed out waiting for ajax calls to start
    When the user removes the public link of file "lorem.txt" using the webUI                                             # WebUISharingContext::theUserRemovesThePublicLinkOfFileUsingTheWebui()
      │ INFORMATION: timed out waiting for ajax calls to start
    Then the public should see an error message "File not found" while accessing last created public link using the webUI # WebUISharingContext::thePublicShouldSeeAnErrorMessageWhileAccessingLastCreatedPublicLinkUsingTheWebui()
SCENARIO RESULT: (pass)

1 scenario (1 passed)
6 steps (6 passed)
0m50.21s (22.45Mb)
runsh: Total 33 scenarios (32 passed, 1 failed)
runsh: Exit code of main run: 0
runsh: Total unexpected failed scenarios throughout the test run:
webUISharingPublic2/shareByPublicLink.feature:151
runsh: There were no unexpected success.
Makefile:142: recipe for target 'test-acceptance-core-webui' failed
make: *** [test-acceptance-core-webui] Error 1
```

--- a/tests/acceptance/run.sh
+++ b/tests/acceptance/run.sh
@@ -509,6 +509,7 @@ function run_behat_tests() {
 		PASSED=true
 		FAILED_SCENARIO_FOUND=false
 		FAILED_SCENARIO_PATHS=`awk '/Failed scenarios:/',0 ${TEST_LOG_FILE} | grep feature`
+
 		for FEATURE_COLORED in ${FAILED_SCENARIO_PATHS}
 			do
 				FAILED_SCENARIO_FOUND=true
@@ -536,10 +537,24 @@ function run_behat_tests() {
 				echo "Rerun failed scenario: ${FAILED_SCENARIO_PATH}"
 				${BEHAT} --colors --strict -c ${BEHAT_YML} -f junit -f pretty ${BEHAT_SUITE_OPTION} --tags ${BEHAT_FILTER_TAGS} ${FAILED_SCENARIO_PATH} -v  2>&1 | tee -a ${TEST_LOG_FILE}
 				BEHAT_EXIT_STATUS=${PIPESTATUS[0]}
-				if [ ${BEHAT_EXIT_STATUS} -ne 0 ]
+				if [ ${BEHAT_EXIT_STATUS} -eq 0 ]
 				then
+					# The scenario was not expected to fail but had failed and is present in the
+					# unexpected_failures list. We've checked the scenario with a re-run and
+					# it passed. So remove it from the unexpected_failures list.
+					for i in "${!UNEXPECTED_FAILED_SCENARIOS[@]}"
+					do
+						if [ "${UNEXPECTED_FAILED_SCENARIOS[i]}" == "$SUITE_SCENARIO" ]
+						then
+							unset "UNEXPECTED_FAILED_SCENARIOS[i]"
+						fi
+					done
+				else
 					echo "webUI test rerun failed with exit status: ${BEHAT_EXIT_STATUS}"
 					PASSED=false
+					# The scenario is not expected to fail but is failing also after the rerun.
+					# Since it is already reported in the unexpected_failures list, there is no
+					# need to touch that again. Continue processing the next scenario to rerun.
 				fi
 			done
 

